GPIO Bit-banging Dshot with GCR Telemetry

joelucid opened the pull request:
This PR implements GPIO bit banging DShot. See https://github.com/betaflight/betaflight/pull/7446 for earlier discussion. Thanks @jflyper for the gpio bit banging implementation!

To enable ``set dshot_bbshot=on``.

The feature will allow almost all F4 and F7 FC's to work with bidirectional dshot out of the box, without reconfiguration of DMA and timers. Given the importance of rpm telemetry to get the best possible flight experience with 4.1 and the quickly approaching feature freeze date I'm submitting this PR early to get reviews started.

I've tested the feature with the NOX, OMNIBUSF4, FURYF4OSD and MATEKF722SE targets with dshot300 and dshot600. Jesc Telemetry 2.0 does not work well with dshot600 in this implementation. I have a fix which will be released when this feature gets integrated. Dshot300 works with jesc Telemetry 2.0.

Performance is about on par with the timer based implementation.

Much of the implementation is straight forward. However there is some timer hackery to achieve the dynamic allocation of bit bang pacing timers and dma streams depending on the other needs of the system. 

For this I add in the target timer config one iotag-less timer for each dmaopt available for each channel of timers 1 and 8 where available. Motors are only pre-initialized to get steady line output while the system initializes. The real initialization then happens when all other timers and dma streams have been allocated, allowing the bbshot driver to pick ones that aren't in use.

I'm sure what I've done here isn't ideal - the strong binding between timers and pins just makes it hard to reflect available timers and dmas. This is just the best I could come up with within the constraints of the current architecture. In particular I'm guessing changes will be required to have this work with universal targets.

mikeller commented on src/main/drivers/dshot_bb.c:
```diff
@@ -0,0 +1,697 @@
+    UNUSED(idlePulse);
+    UNUSED(useUnsyncedPwm);
+
+    dbgPinInit();
```
What is this used for?

joelucid replied:
I defer to @jflyper

jflyper replied:
These provide minimal latency timing marker output used for optimizing raw I/O codes. We can strip them now, but I would like to keep them for a while, until all platform support is done (H7 and possibly G4).

mikeller replied:
They seem to be generically useful for debugging timing issues - I think they should be extracted into separate files, in a separate pull request.

Or at least they need to be commented properly, so that if we forget to remove them in the end, the next person editing the code will not have to spend hours guessing and testing if this functionality is actually required for bitbanged Dshot to work.

mikeller commented on src/main/drivers/dshot_bb_impl.h:
```diff
@@ -0,0 +1,228 @@
+#define DEBUG_COUNT_INTERRUPT
+#define DEBUG_MONITOR_PACER
+
+#define BBSHOT_TELEMETRY_OVER_SAMPLE 3
```
`BITBANGING_DSHOT_TELEMETRY_OVER_SAMPLE`?

joelucid replied:
done

mikeller commented on src/main/drivers/pwm_output_dshot_shared.h:
```diff
@@ -77,5 +74,6 @@ FAST_CODE void pwmDshotSetDirectionOutput(
 
 bool pwmStartDshotMotorUpdate(void);
 
+
```
:space_invader:

joelucid replied:
done

mikeller commented on src/main/drivers/resource.c:
```diff
@@ -106,5 +106,6 @@ const char * const ownerNames[OWNER_TOTAL_COUNT] = {
     "PULLUP",
-    "PULLDOWN"
+    "PULLDOWN",
+    "BBSHOT",
```
`BITBANGING_DSHOT`

joelucid replied:
done

mikeller commented on src/main/drivers/resource.h:
```diff
@@ -105,6 +105,7 @@ typedef enum {
     OWNER_BARO_XCLR,
     OWNER_PULLUP,
     OWNER_PULLDOWN,
+    OWNER_BBSHOT,
```
`OWNER_BITBANGING_DSHOT`

joelucid replied:
done

mikeller commented on src/main/drivers/timer.h:
```diff
@@ -82,6 +82,7 @@ typedef enum {
     TIM_USE_TRANSPONDER    = 0x20,
     TIM_USE_BEEPER         = 0x40,
     TIM_USE_CAMERA_CONTROL = 0x80,
+    TIM_USE_BBSHOT         = 0x100, // BB-Dshot pacer
```
`TIM_USE_BITBANGING_DSHOT_PACER`, and then get rid of the comment? ;-)

joelucid replied:
That flag I could eliminate after moving to a separate timerHardware array for bb

mikeller commented on src/main/drivers/timer_common.c:
```diff
@@ -158,5 +158,19 @@ ioTag_t timerioTagGetByUsage(timerUsageFlag_e usageFlag, uint8_t index)
+
+const timerHardware_t *timerGetByUsage(timerUsageFlag_e usageFlag, uint8_t index)
+{
+    uint8_t currentIndex = 0;
```
This will have to be hardcoded to return `NULL` if `USE_UNIFIED_TARGET` is not defined.

joelucid replied:
Also thrown out now.

mikeller commented on src/main/drivers/timer_def.h:
```diff
@@ -1004,4 +1014,41 @@
+#endif
+
+
+#ifdef USE_BBSHOT
```
This seems to be the wrong file for this - it is used to define the MCU hardware. Bitbanging Dshot specific defines should probably go into a file of their own.

joelucid replied:
Yup - this is now in dshot_bitbang.c

mikeller commented on src/main/target/AG3X/target.c:
```diff
@@ -60,4 +60,6 @@ const timerHardware_t timerHardware[USABLE_TIMER_CHANNEL_COUNT] = {
     // I2C2
     DEF_TIM(TIM2,  CH3, PB10, TIM_USE_NONE,   0, 0), // SCL2   D(1,1)
     DEF_TIM(TIM2,  CH4, PB11, TIM_USE_NONE,   0, 0), // SDA2   D(1,7) D(1,6)
+    BBSHOT_PACER_TIMER_CHANNELS
```
Why are these added twice?

Also, instead of modifying all target definitions, this should really be done by making the code consuming `timerHardware` aware of bitbanging Dshot.

Also, as remarked earlier, this approach will not work for Unified Targets, so we need to look for a different approach.

joelucid replied:
Yeah this was addressed with the separate bbTimerHardware array.

mikeller commented on src/main/target/AG3X/target.h:
```diff
@@ -170,5 +170,5 @@
 
-#define USABLE_TIMER_CHANNEL_COUNT 15
-#define USED_TIMERS ( TIM_N(1) | TIM_N(2) | TIM_N(3) | TIM_N(4) | TIM_N(5) | TIM_N(8) | TIM_N(9) | TIM_N(12))
+#define USABLE_TIMER_CHANNEL_COUNT ((15 + BBSHOT_PACER_COUNT) + BBSHOT_PACER_COUNT)
```
Same as above, should be handled in the code consuming `USABLE_TIMER_CHANNEL_COUNT`.

jflyper commented:
The file name `telemetry_decode.[ch]` sounds too generic. `dshot_bb_gcr.[ch]`?

joelucid commented on src/main/cli/settings.c:
```diff
@@ -755,6 +755,9 @@ const clivalue_t valueTable[] = {
     { "dshot_bidir",            VAR_UINT8  | MASTER_VALUE | MODE_LOOKUP, .config.lookup = { TABLE_OFF_ON }, PG_MOTOR_CONFIG, offsetof(motorConfig_t, dev.useDshotTelemetry) },
 #endif
+#ifdef USE_BBSHOT
+    { "dshot_bbshot",               VAR_UINT8  | MASTER_VALUE | MODE_LOOKUP, .config.lookup = { TABLE_OFF_ON }, PG_MOTOR_CONFIG, offsetof(motorConfig_t, dev.useBBShot) },
```
Yeah - I've renamed it to dshot_bitbang and useDshotBitbang.

joelucid commented on src/main/drivers/dshot.c:
```diff
@@ -50,6 +50,8 @@
 
 #include "rx/rx.h"
 
+
```
done

joelucid commented on src/main/drivers/dshot_bb.c:
```diff
@@ -0,0 +1,697 @@
+
+#include "platform.h"
+
+#ifdef USE_BBSHOT
```
I changed to USE_DSHOT_BITBANG to keep it just a tad shorter.

joelucid commented on src/main/drivers/dshot_bb.c:
```diff
@@ -0,0 +1,697 @@
+    }
+}
+
+static void bbAllocDMA(bbPort_t *bbPort)
```
done

joelucid commented on src/main/drivers/dshot_bb.c:
```diff
@@ -0,0 +1,697 @@
+
+bool bbMotorConfig(IO_t io, uint8_t motorIndex, motorPwmProtocolTypes_e pwmProtocolType, uint8_t output)
+{
+    UNUSED(output); // Not supported
```
definitely!

joelucid replied:
Actually it was still used so I removed the UNUSED

joelucid commented on src/main/drivers/dshot_bb.c:
```diff
@@ -0,0 +1,697 @@
+    const timerHardware_t *timer;
+    bool usedTimerChannels[32][CC_CHANNELS_PER_TIMER] = {0};
+
+    for (int bbPortIndex = 0; bbPortIndex < MAX_SUPPORTED_MOTOR_PORTS; bbPortIndex++) {
```
I did that for targets like this:
```
src/main/target//OMNIBUSF7/target.c:    DEF_TIM(TIM8, CH3, PC8, TIM_USE_CAMERA_CONTROL, 0, 1), // USED FOR CAMERA CONTROL
src/main/target//OMNIBUSF7/target.c:    DEF_TIM(TIM1, CH3, PE13, TIM_USE_NONE,          0, 1 ), // RC1 / PPM, unusable
src/main/target//OMNIBUSF7/target.c:    DEF_TIM(TIM1, CH1, PE9,  TIM_USE_MOTOR,         0, 2 ), // M3
src/main/target//OMNIBUSF7/target.c:    DEF_TIM(TIM1, CH2, PE11, TIM_USE_MOTOR,         0, 1 ), // M4
src/main/target//OMNIBUSF7/target.c:    DEF_TIM(TIM8, CH1, PC6,  TIM_USE_NONE,          0, 0 ), // UART6_TX
src/main/target//OMNIBUSF7/target.c:    DEF_TIM(TIM8, CH2, PC7,  TIM_USE_NONE,          0, 1 ), // UART6_RX
```

As you see both TIM1 and TIM8 are made available in this target, but for uses that are likely fairly rare (at least both together). Picking dynamically what's still unused allows these targets to work in the vast majority of all cases.

mikeller replied:
Ah, so this only applies to targets that use TIM1 and TIM8 as pin based timers? In this case the potential for conflict is much smaller, and doing this dynamically should be fine.

What happens if there aren't enough timers available, how is this indicated to the user?

joelucid replied:
<img width="755" alt="image" src="https://user-images.githubusercontent.com/18134492/64071559-0518cb00-cc7d-11e9-89d7-06f008b01d5d.png">

Yeah - we can only use timer 1 and timer 8 as gpio bitbang pacers since they are the only  timers that trigger dma streams on dma controller 2 which is the only dma controller that can access APB2 where the gpio IO registers reside (see picture, it's the same on stm32f7).

If no pacing timer can be found the quad will not arm since there's no telemetry available.

mikeller replied:
> If no pacing timer can be found the quad will not arm since there's no telemetry available.

What happens if Dshot telemetry is set to off (while `dshot_bitbang` is `on`)?

Probably better to fail obvious in this case by directly setting arming disabled. For every other function we have, it is simple and obvious that if a resource that is configured as required but not claimed it won't work. With this, since we are automatically allocating resources, it will be hard for the user to determine if this should be working or not.

joelucid replied:
Yes true and good point.

joelucid replied:
done

joelucid commented:
> The file name `telemetry_decode.[ch]` sounds too generic. `dshot_bb_gcr.[ch]`?

done

mikeller commented on src/main/drivers/dshot_bitbang.c:
```diff
@@ -0,0 +1,710 @@
+#endif
+
+    for (int motorIndex = 0; motorIndex < MAX_SUPPORTED_MOTORS && motorIndex < motorCount; motorIndex++) {
+        const timerHardware_t *timerHardware = timerGetByTag(motorConfig->ioTags[motorIndex]);
```
Is this timer actually available for other uses when bitbanging Dshot is used? If not, then I this should use `timerAllocate(motorConfig->ioTags[motorIndex], OWNER_DSHOT_BITBANG, motorIndex)` to mark the timer as used.

joelucid replied:
These timers are available for other uses. They are only accessed here to check whether the signal should be inverted.

ledvinap replied:
Isn't pin inversion hardwired into Timer PWM? There should be no inversion necessary for bitbang ...

mikeller commented on src/main/target/common_defaults_post.h:
```diff
@@ -659,5 +659,14 @@
+
+#ifdef USE_DSHOT_BITBANG
+#ifndef ENABLE_BBSHOT
+#define ENABLE_BBSHOT false
```
This seems to be unused now.

joelucid replied:
done

mikeller commented on src/main/target/common_defaults_post.h:
```diff
@@ -659,5 +659,14 @@
 #define MAX_SUPPORTED_MOTORS 8
 #endif
+#ifndef MAX_SUPPORTED_MOTOR_PORTS
+#define MAX_SUPPORTED_MOTOR_PORTS 4 // Max direct dshot port groups, limited by number of usable timer (TIM1 and TIM8) x number of channels per timer (4), 3 is enough to cover motor pins on GPIOA, GPIOB and GPIOC.
```
Does it make sense to have this per-target configurable, or should this just be a fixed define? (Per - target configurable settings will become obsolete with Unified Targets anyway.)

joelucid replied:
This depends on the maximum number of ports that might be used on a given architecture. So with unified targets it would be specified by architecture and still makes sense. 4 should cover all cases for now.

mikeller replied:
I'd move that out of here and into the bitbanged Dshot units - no point having one more thing that we have to support being not what we expect it to be.

mikeller commented on src/main/drivers/motor.c:
```diff
@@ -206,7 +217,16 @@ void motorDevInit(const motorDevConfig_t *motorConfig, uint16_t idlePulse, uint8
+#ifdef USE_DSHOT_BITBANG
+        if (motorConfig->useDshotBitbang == DSHOT_BITBANG_ON ||
+            (motorConfig->useDshotBitbang == DSHOT_BITBANG_AUTO && motorConfig->useDshotTelemetry)) {
+            motorDevice = bbshotDevInit(motorConfig, idlePulse, motorCount, useUnsyncedPwm);
```
`dshotBitbangDevInit()` matches the naming below.

joelucid replied:
done

mikeller commented on src/main/drivers/dshot_bitbang_decode.c:
```diff
@@ -0,0 +1,133 @@
+        }
+    }
+
+    if(*p & mask) {
```
:space_invader:

etracer65 replied:
Some documentation in this section would be really nice. The code is very obscure and not many people are going to understand it later.

joelucid replied:
I've documented the whole function now.

mikeller commented on src/main/drivers/dshot_bitbang_decode.c:
```diff
@@ -0,0 +1,133 @@
+    sequence[sequenceIndex++] = p - buffer;
+#endif
+
+    while ( endP > p ) {
```
:space_invader:

joelucid replied:
done

mikeller commented on src/main/drivers/dshot_bitbang_impl.h:
```diff
@@ -0,0 +1,228 @@
+
+extern FAST_RAM_ZERO_INIT bbMotor_t bbMotors[MAX_SUPPORTED_MOTORS];
+
+extern FAST_RAM_ZERO_INIT bbPort_t *gpioToMotorPort[10]; // GPIO group to bbPort mapping
```
Can we get rid of the magic number here?

But it would probably be cleaner to change this to only use `bbPorts`, and iterate through them to find the one with the right port index.

joelucid replied:
done

mikeller commented:
This is starting to look good - the separation of the bitbanged timer array together with the dynamic allocation will make this work with Unified Targets.

We should look into making the pacer timers visible in CLI, probably as part of `timer show`. This could be as simple as making `timer show` aware of `bbPorts` when iterating over the timer list.

mikeller commented on src/main/drivers/itimer.c:
```diff
@@ -0,0 +1,17 @@
+
+// Generic interval timer service
```
This seems to be a stray file (also, it is picked up by the sanity checks for not including `platform.h`).

joelucid replied:
removed

mikeller commented on src/main/drivers/dshot_bitbang.c:
```diff
@@ -0,0 +1,710 @@
+
+
+const timerHardware_t bbTimerHardware[] = {
+#if defined(STM32F411xE) || defined(STM32F40_41xxx) || defined(STM32F7)
```
The conditionals need to be cleaned up, as this is now breaking for the F446.

joelucid replied:
done

joelucid commented on src/main/drivers/dshot_bitbang_decode.c:
```diff
@@ -0,0 +1,133 @@
+#define MIN_VALID_BBSAMPLES ((21 - 2) * 3)
+#define MAX_VALID_BBSAMPLES ((21 + 2) * 3)
+
+#ifdef DEBUG_BBDECODE
```
I've added a comment.

joelucid commented on src/main/drivers/dshot_bitbang_decode.c:
```diff
@@ -0,0 +1,133 @@
+
+FAST_CODE uint32_t decode_bb( uint16_t buffer[], uint32_t count, uint32_t mask)
+{
+//    printf("start\n");
```
done

joelucid commented on src/main/drivers/pwm_output.c:
```diff
@@ -37,7 +37,7 @@
 FAST_RAM_ZERO_INIT pwmOutputPort_t motors[MAX_SUPPORTED_MOTORS];
 
-static void pwmOCConfig(TIM_TypeDef *tim, uint8_t channel, uint16_t value, uint8_t output)
+void pwmOCConfig(TIM_TypeDef *tim, uint8_t channel, uint16_t value, uint8_t output)
```
@jflyper ?

jflyper replied:
It's probably a side effect of motor code refactoring which my original commit was based on.
We can add back the `static` safely, I suppose.

joelucid replied:
added back in

joelucid commented on src/main/drivers/dshot_bitbang.c:
```diff
@@ -0,0 +1,710 @@
+
+motorDevice_t *bbshotDevInit(const motorDevConfig_t *motorConfig, uint16_t idlePulse, uint8_t count, bool useUnsyncedPwm)
+{
+    UNUSED(idlePulse);
```
done

joelucid commented:
> This is starting to look good - the separation of the bitbanged timer array together with the dynamic allocation will make this work with Unified Targets.
> 
> We should look into making the pacer timers visible in CLI, probably as part of `timer show`. This could be as simple as making `timer show` aware of `bbPorts` when iterating over the timer list.

I added the bitbang timers to ``timer show`` now.

mikeller commented:
> I added the bitbang timers to `timer show` now.

Nice! One other thought that I had was that it would be nice to add a 'Bitbanged Dshot: enabled / disabled' to `dshot_telemetry_info` - should be trivial.

mikeller commented on src/main/drivers/dshot_bitbang.c:
```diff
@@ -0,0 +1,724 @@
+    return &bbDevice;
+}
+
+typedef struct dbgPin_s {
```
I am still intrigued by what this does? @jflyper?
Looks like some sort of debug facility, and in this case it should probably be extracted into separate files, and put into a separate pull request.

jflyper replied:
Yes, it's a low latency timing marker facility mainly used for precise timing measurements. (Pinio can serve for casual measurements.) Lemme see how I can generalize this and make them into an independent facility.

mikeller commented:
Flight tested with STM32F405 and STM32F411 Unified Targets driving JESC ESCs with Dshot300, confirmed working fine.

joelucid commented:
Please don't merge yet. Seen numerous issues this afternoon: corrupt rpm values making it to the filter once in a while, soft serial telemetry not working right due to excessive cpu at 4k4k on F411 and a couple of weird flights problems.

joelucid commented:
I've found and fixed the issue with soft-serial. Still got some corrupted packets when logging rpm data so I added more error detection to the decoding routine. Dshot600 on my test quad still has 2-3 corrupted packets that make it through checksum and error detection, but only on one motor.

I think this is getting good to go.

joelucid commented:
@jflyper, I changed some of the DMA parameters: I now use word alignment on memory, half word on peripheral and enabled the fifo. This should cut the used memory bandwidth in half. Would appreciate if you could look it over and check that you don't see any issues. Doesn't seem to cause any changes in performance.

ledvinap commented on src/main/drivers/dshot_bitbang.c:
```diff
@@ -0,0 +1,729 @@
+#endif
+        timeUs_t currentUs = micros();
+        // don't send while telemetry frames might still be incoming
+        if (currentUs - lastSendUs < 40 + 2 * dshotFrameUs) {
```
(`cmpTimeUs()`)

joelucid replied:
done

ledvinap commented on src/main/drivers/dshot_bitbang.c:
```diff
@@ -0,0 +1,729 @@
+
+void dbgPinHi(int index)
+{
+    if ((unsigned)index > ARRAYLEN(dbgPins)) {
```
(off-by-one error, `>=` is correct)

joelucid replied:
done

ledvinap commented on src/main/drivers/dshot_bitbang_decode.c:
```diff
@@ -0,0 +1,290 @@
+#define BITBAND_SRAM_BASE  0x22000000
+#define BITBAND_SRAM(a,b) ((BITBAND_SRAM_BASE + ((a-BITBAND_SRAM_REF)<<5) + (b<<2)))  // Convert SRAM address
+
+typedef struct bitBandWord_s {
```
bitBandWord_t is unaligned this way (and uint8_t). Using uint32_t for value will probably produce faster code

joelucid replied:
The critical part of the code looked good - the code executed for each bit. I changed to uint32_t away.

ledvinap commented on src/main/drivers/dshot_bitbang_decode.c:
```diff
@@ -0,0 +1,290 @@
+/* Bit band SRAM definitions */
+#define BITBAND_SRAM_REF   0x20000000
+#define BITBAND_SRAM_BASE  0x22000000
+#define BITBAND_SRAM(a,b) ((BITBAND_SRAM_BASE + ((a-BITBAND_SRAM_REF)<<5) + (b<<2)))  // Convert SRAM address
```
`()` around macro arguments ...

joelucid replied:
done

ledvinap commented on src/main/drivers/dshot_bitbang_decode.c:
```diff
@@ -0,0 +1,290 @@
+/* Bit band SRAM definitions */
+#define BITBAND_SRAM_REF   0x20000000
+#define BITBAND_SRAM_BASE  0x22000000
+#define BITBAND_SRAM(a,b) ((BITBAND_SRAM_BASE + ((a-BITBAND_SRAM_REF)<<5) + (b<<2)))  // Convert SRAM address
```
Also ,... isn't `a` 16 bit wide (*64 to get bitband address)?

joelucid replied:
![image](https://user-images.githubusercontent.com/18134492/64168608-34316700-ce4c-11e9-8474-2aa0ca548120.png)

ledvinap replied:
Yes, you are right ... `a` is already multiplied ...

ledvinap commented on src/main/drivers/dshot_bitbang_decode.c:
```diff
@@ -0,0 +1,290 @@
+    bitBandWord_t* endP = p + (count - MIN_VALID_BBSAMPLES);
+
+    // Eliminate leading high signal level by looking for first zero bit in data stream.
+    // Manual loop unrolling and branch hinting to produce faster code.
```
(BTW: did you reviewed generated code? unrolling may not be worth it)

joelucid replied:
Yeah it's definitely worth it. I was hoping for 2 instructions per sample but unfortunately compiler generates 3 (doesn't manage to use CJZ, CJNZ). Considering there may be 9 consecutive samples of the same level not having to do the p increase, the p < endP comparison and the branch for each bit does make a big difference.

ledvinap replied:
(Using uint32_t may help a bit ... )

joelucid replied:
Nah - it's just ldrw vs ldrb or something similar. arm does widening in the same instruction, no cycle cost.

joelucid replied:
On the other hand when searching the first zero we could do all bits on one port together and  look at the not bit banded buffer using 32 bit to check two data points in one step. That would help a bit. Someday ...

ledvinap replied:
Yes, but each ESC may reply with different delay, so it will handle only common part ...

joelucid replied:
Which is the main part since we don't detect the response or predict when it'll arrive. But icing on the cake.

ledvinap commented on src/main/drivers/dshot_bitbang_impl.h:
```diff
@@ -0,0 +1,229 @@
+#define MOTOR_PROSHOT1000_BIT_PER_SYMBOL   4
+
+#define MOTOR_DSHOT_STATE_PER_SYMBOL       3  // Initial high, 0/1, low
+#define MOTOR_PROSHOT1000_STATE_PER_SYMBOL 32 // 8 preamble + 16 + 8 postamble
```
Proshot is terribly wasteful when using bitbangind, it may be better to drop it)

joelucid replied:
I agree

mikeller replied:
As far as I know, there is no ESC hardware that does ProShot1000 and **with**  RPM feedback (yet). Can we simplify this by disabling bitbanged Dshot if ProShot is selected, and then stripping the ProShot code out of this?

joelucid replied:
blheli32 does proshot1000 with rpm telemetry.

ledvinap replied:
There isn't problem with proshot in general. Proshot is good protocol if resources are spent per edge (timer case). But using it with bitbanging is overkill, it is probably OK to force user to avoid this combination.

mikeller replied:
Ok, I should have said 'does well'. @sskaug stated that ProShot1000 has got timing issues on BLHeli_32 hardware.

mikeller replied:
@ledvinap: That's exactly what I suggested - but in this case we should strip the ProShot1000 code from the bitbanged Dshot code as it will only rot anyway if it cannot be used.

joelucid replied:
I removed the 'proshot1000' code and changed 'dshot_bitbang = auto' not to select bitbang with 'proshot1000'. Additionally I changed post config validation to 'set dshot_bitbang = auto' if it's on and proshot1000 is selected.

joelucid commented on src/main/drivers/dshot.h:
```diff
@@ -20,10 +20,34 @@
+//#define DEBUG_BBDECODE
+
+#ifdef USE_DSHOT_TELEMETRY_STATS
+#define DSHOT_TELEMETRY_QUALITY_WINDOW 1       // capture a rolling 1 second of packet stats
```
This was just moved between files. Let's focus on the changes in light of 4.1 RC coming.

joelucid commented on src/main/drivers/dshot.h:
```diff
@@ -20,10 +20,34 @@
+typedef struct dshotTelemetryQuality_s {
+    uint32_t packetCountSum;
+    uint32_t invalidCountSum;
+    uint32_t packetCountArray[DSHOT_TELEMETRY_QUALITY_BUCKET_COUNT];
```
This was just moved between files. Let's focus on the changes in light of 4.1 RC coming.

joelucid commented on src/main/drivers/dshot_bitbang.c:
```diff
@@ -0,0 +1,729 @@
+    }
+}
+
+void bbOutputDataSet(uint32_t *buffer, int portNumber, uint16_t value, bool inverted)
```
Yes

joelucid commented on src/main/drivers/dshot_bitbang.c:
```diff
@@ -0,0 +1,729 @@
+        uint16_t bufferSize = (pwmProtocolType == PWM_TYPE_PROSHOT1000) ? MOTOR_PROSHOT1000_BUFFER_SIZE : MOTOR_DSHOT_BUFFER_SIZE;
+
+        bbPort->portOutputCount = bufferSize;
+        bbPort->portOutputBuffer = &bbOutputBuffer[(bbPort - bbPorts) * bufferSize];
```
@jflyper could you comment?

ledvinap replied:
Even better, use 2d array ... [ports][bytes per port]
